gym_adserver/wrappers/compare_agents.py

```diff
@@ -0,0 +1,56 @@
+import argparse
+
+import gym
+from gym import logger, spaces
+from joblib import Parallel, delayed
+
+from gym_adserver.agents.epsilon_greedy_agent import EpsilonGreedyAgent
+from gym_adserver.agents.random_agent import RandomAgent
+from gym_adserver.agents.softmax_agent import SoftmaxAgent
+from gym_adserver.agents.ucb1_agent import UCB1Agent
+
+
+def setup_environment(env_name, num_ads, time_series_frequency):
+    env = gym.make(env_name, num_ads=num_ads, time_series_frequency=time_series_frequency)
+    env.seed(args.seed)
+    return env
+
+if __name__ == '__main__':
+    parser = argparse.ArgumentParser()
+    parser.add_argument('--env', default='AdServer-v0')
+    parser.add_argument('--num_ads', type=int, default=10)
+    parser.add_argument('--impressions', type=int, default=10000)
+    parser.add_argument('--seed', type=int, default=0)
+    parser.add_argument('--output_file', default=None)
+    args = parser.parse_args()
+
+    logger.set_level(logger.INFO)
+
+    time_series_frequency = args.impressions // 10
+    action_space = spaces.Discrete(args.num_ads)
+    max_impressions = args.impressions
+    seed = args.seed
+
+    # Define your agents here
+    agents = [
+        RandomAgent(action_space=action_space),
+        SoftmaxAgent(seed=seed, beta=5, max_impressions=max_impressions),
+        UCB1Agent(action_space=action_space, seed=args.seed, c=2, max_impressions=max_impressions),
+        EpsilonGreedyAgent(seed=seed, epsilon=0.1),
+    ]
+
+    # Simulation loop
+    envs = []
+    for agent in agents:
+        reward = 0
+        env = setup_environment(env_name=args.env, num_ads=args.num_ads, time_series_frequency=time_series_frequency)
+        observation = env.reset(agent.name)
+        envs.append(env)
+        for i in range(args.impressions):
+            # Action/Feedback
+            ad_index = agent.act(observation=observation, reward=reward, done=False)
+            observation, reward, _, _ = env.step(ad_index)
+
+    # Render result for each agent (NOTE: close all to quit)
+    parallel = Parallel(n_jobs=-1)
```

Comment:
I like the parallel simulation, but I'm not sure about your intention here. Are you maybe using Parallel to keep open all the N charts simultaneously?

I see a couple of (non-blocking) issues here, to fix later:

- Logger is not propagated automatically
- All the N threads are blocked until the plots are closed

Maybe in future we could render all the results in a single window (with a different layout), so the second point could be indirectly solved.

Reply:
I agree, merging it all in 1 plot would be better. But it could grow quite big so not sure how to deal with that other than looking for a solution with a scrollable window.
Feel free to merge if you want.